### PR TITLE
Better settings

### PR DESCRIPTION
**Pull request by niosus**
An overhaul of settings:
- deprecated `use_project_specific_settings` in favor of specifying settings under `settings` in `*.sublime-project` file. See the `easy_clang_complete.sublime-project` in this repo for an example. This fixes issues #64 and #65 
- rewritten settings don't need manual typing anymore and are read from class `__dict__` instead.

### CODE REVIEW

File: plugin/completion/bin_complete.py [hunk 1]
```diff
@@ -150,29 +150,21 @@ def init(self, view, settings):
...
-            for include in includes:
-                clang_flags.append('-I "{}"'.format(include))
-
-            if settings.search_clang_complete and self.flags_manager:
-                log.debug(" flags_manager loaded")
-                custom_flags = self.flags_manager.get_flags(
-                    separate_includes=True)
-                clang_flags += custom_flags
+        # init needed variables from plugin settings as project settings
+        # are either not used or invalid
```
Comment (rchl) [inline plugin/completion/bin_complete.py]: Not sure I understand this comment. Why are project settings not used or invalid at this point?

Comment (rchl) [inline plugin/completion/bin_complete.py]: Now I see that it's a comment from old code and probably doesn't make sense now.

File: plugin/completion/lib_complete.py [hunk 2]
```diff
@@ -163,33 +163,22 @@ def init(self, view, settings):
...
-
-            if settings.search_clang_complete and self.flags_manager:
-                log.debug(" flags_manager loaded")
-                custom_flags = self.flags_manager.get_flags(
-                    separate_includes=False)
-                clang_flags += custom_flags
+        # add std flag to all flags
+        clang_flags.append(std_flag)
+        # this means that project specific settings are either not used or
+        # invalid, so we still need to initialize from settings
```
Comment (rchl) [inline plugin/completion/lib_complete.py]: Comment doesn't make sense now.

File: plugin/plugin_settings.py [hunk 3]
```diff
@@ -16,52 +16,59 @@
 log.debug(" reloading module")
 
 
+class SettingsEnum:
+    include_dirs = None
```
Comment (rchl) [inline plugin/plugin_settings.py]: Could sort properties by name.

File: plugin/plugin_settings.py [hunk 4]
```diff
@@ -139,35 +118,35 @@ def load_settings(self):
...
-            log.error(" failed to read clang flags from project settings.")
-            log.error(" error is: %s.", e)
-            return None
+        log.debug(" overriding settings by project ones if needed:")
+        settings_handle = sublime.active_window().active_view().settings()
+        self.__load_vars_from_settings(settings_handle)
+        log.debug(" done.")
+
+    def __load_vars_from_settings(self, settings_handle):
+        for key, value in SettingsEnum.__dict__.items():
```
Comment (rchl) [inline plugin/plugin_settings.py]: I believe using `_` would satisfy linter.

File: plugin/plugin_settings.py [hunk 5]
```diff
@@ -139,35 +118,35 @@ def load_settings(self):
...
+            if key.startswith('__') or callable(key):
```
Comment (rchl) [inline plugin/plugin_settings.py]: Would maybe make sense to just use normal dict and iterate it using normal loop. Then you don't have to care about any existing properties.

Define dict like a static global:

SETTINGS_DICT = {...}

File: plugin/plugin_settings.py [hunk 6]
```diff
@@ -139,35 +118,35 @@ def load_settings(self):
...
+                continue
+            val = settings_handle.get(key)
+            if val is not None:
+                value = val
```
Comment (rchl) [inline plugin/plugin_settings.py]: Why creating new variable?

**Comment by niosus** [comment]
@rchl thanks for all the reviews. I have pushed a working version, but haven't done any cleanup yet. As it was late night for me (3 am) I decided I will do the pass through when my head is fresh :smile: 

All the things you say make sense, so I will incorporate them probably later today. Thanks!

**Pull request merged**